### PR TITLE
164796891 Users Should Be Able To Create Articles

### PR DESCRIPTION
**Pull request by chikeozulumba**
#### Description
- Create an API for creating articles
#### Type of change
- New feature (non-breaking change that adds functionality)
#### How Has This Been Tested?
- Unit test
- Integration test
#### Checklist:
- [x] Style guide
#### PT
[#164796891](https://www.pivotaltracker.com/story/show/164796891)

### CODE REVIEW

File: src/routers/index.js [hunk 1]
```diff
@@ -1,9 +1,27 @@
...
+ * @function
+ * @memberof module:Express.Router
+ * @inner
+ * @param {function} AddArticles - Express path
+ * @param {function} DeleteArticle - Express path
+ * @param {function} UpdateArticle - Express path
+ * @returns Response Object
+ */
+router
+  .route('/api/v1/articles/:id?')
```
Comment (Simplemart17) [inline src/routers/index.js]: You might want to add the base route to the server file.

File: src/controllers/article.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * @name AddArticles
+ * @description This is the method for inserting articles
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @returns {int} Returns the inserted article after success
+ */
+export const AddArticles = async (req, res) => {
+  const validate = await validateArticle(req.body);
```
Comment (Temmyogunbo) [inline src/controllers/article.js]: Let us destructure use ES6 destructing here.i.e 
`const {body} = req;`

File: src/server.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ export const startServer = port => new Promise((resolve, reject) => {
   app.use(bodyParser.urlencoded({ extended: false }));
   app.use(cors());
 
-  app.use('/api/v1', indexRouter);
+  app.use(router);
```
Comment (Temmyogunbo) [inline src/server.js]: Implement @Simplemart17 feedback so we have something like this:
`app.use('/api/v1', router)`

File: tests/unit/article.test.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import 'chai/register-should';
+import { expect } from 'chai';
+import { validateArticle } from '../../src/utils/article';
+
+const ARTICLE = {
```
Comment (Temmyogunbo) [inline tests/unit/article.test.js]: Let us create a folder for our mock data.

File: tests/unit/article.test.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  title: 'Andela is cool 7888',
+  userId: '979eaa2e-5b8f-4103-8192-4639afae2ba8',
+  description: 'Lorem ipsum dolor sit amet, sit ut dolor alterum, sed malis referrentur cu. Aperiam fabulas eos ea. Sea mazim senserit tincidunt te.',
+  body: 'Lorem ipsum dolor sit amet, sit ut dolor alterum, sed malis referrentur cu. Aperiam fabulas eos ea. Sea mazim senserit tincidunt te. Mei volutpat delicatissimi ut, id mollis alienum argumentum has, semper efficiendi sed ea. Ius decore consul forensibus ne, enim verear corpora sit ut. Usu eu possit equidem menandri, quo et noster officiis iracundia.',
+  imageUrl: 'https://picsum.photos/200/300',
+  tags: ['hello', 'async', 'await']
+};
+describe('validateArticle()', () => {
+  it('should return true if the validation passes', async (done) => {
```
Comment (Temmyogunbo) [inline tests/unit/article.test.js]: Why are you using async with thenables? Also, why the done callback?

File: src/controllers/article.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  if (validate.fails()) return res.status(400).json(validate.errors.all());
+  req.body.slug = slug(req.body.title);
```
Comment (Temmyogunbo) [inline src/controllers/article.js]: Let us not mutate.

File: tests/integration/article.test.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+describe('POST /api/v1/articles', () => {
+  it('Should return status: 201', (done) => {
+    agent.post('/api/v1/articles')
+      .send(article)
+      .end((_err, res) => {
+        expect(res).to.have.status(201);
+        expect(res.body).should.be.an('object');
+        expect(res.body).to.have.property('title');
+        expect(res.body.title).to.be.a('string');
+        expect(res.body.tags).to.be.a('array');
```
Comment (Temmyogunbo) [inline tests/integration/article.test.js]: Just so I understand, the controller logic returns article created with slug. We should be testing the value not the property. For instance,
`        expect(res.body.title).to.equal(article.title);
`

File: tests/integration/server.test.js [hunk 8]
```diff
@@ -16,7 +16,7 @@ describe('Root directory Test', () => {
   });
 
   it('Should return status: 200', (done) => {
-    agent.get('/api/v1').end((err, res) => {
+    agent.get('/').end((_err, res) => {
```
Comment (Temmyogunbo) [inline tests/integration/server.test.js]: why the underscore?

File: tests/unit/article.test.js (hunk 5 above)
Comment (chikeozulumba) [reply tests/unit/article.test.js]: @Temmyogunbo Will implement your feedback

File: src/controllers/article.js (hunk 6 above)
Comment (chikeozulumba) [reply src/controllers/article.js]: @Temmyogunbo do I create another object, with the updated article?

File: tests/integration/article.test.js (hunk 7 above)
Comment (chikeozulumba) [reply tests/integration/article.test.js]: @Temmyogunbo Thanks, will implement

File: tests/integration/server.test.js (hunk 8 above)
Comment (chikeozulumba) [reply tests/integration/server.test.js]: @Temmyogunbo Preventing eslint from flagging it as a syntax error

File: tests/unit/article.test.js (hunk 4 above)
Comment (chikeozulumba) [reply tests/unit/article.test.js]: @Temmyogunbo Noted

File: src/server.js (hunk 3 above)
Comment (chikeozulumba) [reply src/server.js]: @Temmyogunbo Noted

File: src/controllers/article.js (hunk 2 above)
Comment (chikeozulumba) [reply src/controllers/article.js]: @Temmyogunbo Noted

File: src/routers/index.js (hunk 1 above)
Comment (chikeozulumba) [reply src/routers/index.js]: @Temmyogunbo Noted

File: src/models/index.js [hunk 9]
```diff
@@ -4,7 +4,7 @@ const Sequelize = require('sequelize');
 
 const basename = path.basename(__filename);
 const env = process.env.NODE_ENV || 'development';
-const config = require('../../config/databaseConfig')[env];
+const config = require('../config/databaseConfig')[env];
```
Comment (ssanusi) [inline src/models/index.js:7]: @chikeozulumba why do we need to the structure of config ?

Comment (chikeozulumba) [reply src/models/index.js:7]: This structure is because i moved the models folder to the root of the src

Comment (ssanusi) [reply src/models/index.js:7]: ok thanks for the clarification

File: src/controllers/article.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+ * @name AddArticles
+ * @description This is the method for inserting articles
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @returns {int} Returns the inserted article after success
+ */
+export const AddArticles = async (req, res) => {
+  const { body } = req;
+  const article = await Article.create({ ...body, slug: slug(body.title) });
+  return res.status(201).json(article);
```
Comment (Temmyogunbo) [inline src/controllers/article.js]: @chikeozulumba what happens when a database failure occurs? Why are we not handling errors?

File: tests/integration/article.test.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+};
+describe('POST /api/v1/articles', () => {
+  it('Should return status: 201', (done) => {
+    agent.post('/api/v1/articles')
+      .send(article)
+      .end((_err, res) => {
+        expect(res).to.have.status(201);
+        const { body } = res;
+        expect(body).should.be.an('object');
+        expect(body).to.have.property('title');
```
Comment (Temmyogunbo) [inline tests/integration/article.test.js]: Please do the same for others.

File: src/routers/index.js [hunk 12]
```diff
@@ -1,9 +1,27 @@
...
+ * @inner
+ * @param {function} AddArticles - Express path
+ * @param {function} DeleteArticle - Express path
+ * @param {function} UpdateArticle - Express path
+ * @returns Response Object
+ */
+router
+  .route('/articles/:id?')
+  .post(AddArticles)
+  .delete(DeleteArticle)
```
Comment (ssanusi) [inline src/routers/index.js:24]: @chikeozulumbaI think to delete article is out of the scope of this ticket

Comment (ssanusi) [inline src/routers/index.js:24]: @chikeozulumbaI think to patch article is out of the scope of this ticket

File: src/controllers/article.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+/**
+ * @name UpdateArticle
+ * @description This is the method for updating an article
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @returns {int} Returns the updated article
+ */
+export const UpdateArticle = async () => true;
```
Comment (ssanusi) [inline src/controllers/article.js:28]: @chikeozulumbaI think to update article is out of the scope of this ticket

File: src/controllers/article.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * @name DeleteArticle
+ * @description This is the method for deleting an article
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @returns {int} Returns true after deleting an article
+ */
+export const DeleteArticle = async () => true;
```
Comment (ssanusi) [inline src/controllers/article.js:37]: @chikeozulumbaI think to delete article is out of the scope of this ticket

**Review by ssanusi — COMMENTED** [review]
@chikeozulumba some implementation of update article, delete article are out of the scope of this ticket

File: src/routers/index.js (hunk 12 above)
Comment (chikeozulumba) [reply src/routers/index.js:24]: @ssanusi  I didn't implement it within the PR, they all have their stories assigned to them, same response to all your comments

**Review by toluola — DISMISSED** [review]
Please Squash your commits to not more than 5 commits

**Pull request merged**